### PR TITLE
Add option to resize with out setting nulls

### PR DESCRIPTION
**Pull request by laithsakka**
### **### ONLY LAST COMMIT BELONGS TO THIS DIFF**
**CLICK ON COMMITS CHOOSE LAST COMMIT AND REVIEW THAT ONLY**

resizing a vector sets all the nulls to not null in the new added portion.
However, in cases such as in vector function adapter that writes the nullity twice.

- Setting nulls cant be avoided because sometimes the vector is pre-allocated with undefined null state.
- Also resize(0) and resize(X) does not work because sometimes we need to keep previously written data. 
to be valid. 

This diff add flag to resize() that can tell the function not to set nulls.

In this diff also I change back the growth rate of capacity to 2. it showed consistent 100ns reduction.
this is because resize is not expensive anymore when done for larger than needed, since the null setting is
gone.

This diff does also fix to the VectorWriter<ArrayProxyT> such that if some data already exists in 
the initial input vector elements, then it does not overwrite it. (it should start writing data 
at the end of the elements vector versus the beginning).

**With nulls**
SimpleProxyPushBack and SimpleGeneral are 2X faster and >2X faster than SimpleOld/
```
After the diff:
============================================================================
VectorBasic                                                284.49ns    3.52M
VectorResizeOptimized                                      218.30ns    4.58M
SimpleProxyWithResize                                      258.74ns    3.86M
SimpleProxyPushBack                                        362.97ns    2.76M
SimpleGeneral                                              312.45ns    3.20M
SimpleOld                                                  664.89ns    1.50M
============================================================================
```

```
Before the diff:
============================================================================
VectorBasic                                                264.58ns    3.78M
VectorResizeOptimized                                      213.93ns    4.67M
SimpleProxyWithResize                                      277.15ns    3.61M
SimpleProxyPushBack                                        732.07ns    1.37M
SimpleGeneral                                              771.50ns    1.30M
SimpleOld                                                  637.80ns    1.57M
============================================================================
```

### CODE REVIEW

**Comment by laithsakka** [comment]
rebase and update numbers

**Comment by laithsakka** [comment]
fix commit history

**Comment by laithsakka** [comment]
format code on the diff stack lol

File: velox/expression/ComplexProxyTypes.h [hunk 1]
```diff
@@ -185,6 +176,15 @@ class ArrayProxy {
     }
   }
 
+  // Should be called by the user (VectorWriter) when writing is done to commit
+  // last item if needed.
+  void finalize() {
+    commitMostRecentChildItem();
+    // Downsize to the actual size used in the underlying vector.
+    // Some writer's logic depend on the previous size to append data.
+    childWriter_->vector().resize(valuesOffset_ + length_);
```
Comment (kgpai) [inline velox/expression/ComplexProxyTypes.h:185]: nit: the indexes after offset + length will all be set as null right ?

**Review by kgpai — APPROVED** [review]

**Comment by mbasmanova** [comment]
@laithsakka Laith, would you extract this change into a separate PR so it can land sooner? It would be nice to document the new 'setNotNull' parameter.

**Comment by laithsakka** [comment]
Landed through a different PR.

**Pull request closed without merging**